.github/workflows/integration_tests.yml

```diff
@@ -166,11 +166,18 @@ jobs:
         run: |
           echo "Autodetecting which tests to run."
           if [[ -n "${{github.event.inputs.test_pull_request}}" ]]; then
-            # If running this manually, diff against main.
-            echo "AUTO_DIFF_PARAM=--auto_diff main" >> $GITHUB_ENV
+            # If running this manually, diff against our common ancestor with main.
+            MERGE_BASE=$(git merge-base HEAD main)
+            echo "::warning ::Auto-diff HEAD..${MERGE_BASE}"
+            git diff --name-only HEAD..${MERGE_BASE}
+            echo "AUTO_DIFF_PARAM=--auto_diff HEAD..${MERGE_BASE}" >> $GITHUB_ENV
           else
             # If running in a PR, diff against the PR's base.
-            echo "AUTO_DIFF_PARAM=--auto_diff ${{github.event.pull_request.base.sha}}" >> $GITHUB_ENV
+            # "git merge-base main branch_name" will give the common ancestor of both branches.
+            MERGE_BASE=$(git merge-base origin/${{github.event.pull_request.head.ref}} origin/${{github.event.pull_request.base.ref}})
+            echo "::warning ::Auto-diff origin/${{github.event.pull_request.head.ref}}..${MERGE_BASE}"
+            git diff --name-only origin/${{github.event.pull_request.head.ref}}..${MERGE_BASE}
+            echo "AUTO_DIFF_PARAM=--auto_diff origin/${{github.event.pull_request.head.ref}}..${MERGE_BASE}" >> $GITHUB_ENV
```

Comment:
I maybe wrong.
But isn't origin/${{github.event.pull_request.head.ref}} equal to HEAD, and ${MERGE_BASE} is equal to origin/${{github.event.pull_request.base.ref}?

Reply:
No, see, this is actually what the issue was.

HEAD is actually a merge commit that's created just for the purpose of the workflow run. For example, HEAD might point to https://github.com/firebase/firebase-cpp-sdk/commit/b069bd7d8edb80d0b3d9137120ab486973f7d463, whereas the latest commit is actually one prior to that - https://github.com/firebase/firebase-cpp-sdk/commit/494ee219290440e3ced7556df989186ffc0a4388.

Meanwhile, origin/main points to the latest head of main in the repo, not the main when the PR was created.

The whole point is that neither of these points to the ref that you'd expect.

Reply:
Here's an example, let me know if this makes sense.

- main: contains commits 111, 222, and 333.
- git checkout main # you are at commit 333
- git checkout -b my_feature_branch
- Then you make changes and commit new commits AAA, BBB, and CCC.
- The commit history for my_feature_branch is now (in reverse order) CCC, BBB, AAA, 333, 222, 111.
- Someone else checks in more commits into main: 444, 555, 666
- Now main is 666, 555, 444, 333, 222, 111 and your branch is CCC, BBB, AAA, 333, 222, 111.
- In order to auto-diff, you need to diff your branch against the common ancestor (in this case 333). This will show the changes that you made in your commits AAA, BBB, and CCC only.
- Now, let's say you merge main into your branch. Your branch now looks like 666, 555, 444, CCC, BBB, AAA, 333, 222, 111.
- But because of the way that merge commits work, the common ancestor is now 666.
- If someone else checks in commits 777 and 888 into main at this point, your common ancestor with main is still 666.
- This will still correctly determine the diff from your branch, even if you merge commits from main into your branch as you go, because diffing your branch against 666 still correctly shows the changes in AAA, BBB, and CCC only.

I realize that was probably entirely unclear - sorry!